### PR TITLE
New Chart: initial stage of Heatmap chart

### PR DESCRIPTION
**Pull request by Golodhros**
## Description

A Heatmap chart based on the code from my blog post for Heart Internet.

tutorial__heatmap_chart

This is still very basic, in the next days I want to add:
* Vertical/Horizontal option
* Dynamic box sizes based on the width/height of the chart
* Date Label formats

## Motivation and Context
Heatmap chart was planned:
https://github.com/eventbrite/britecharts/issues/299

## How Has This Been Tested?
Added tests and demo

## Types of changes

- [ ] Refactor (changes the way we code something without changing its functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: src/charts/grouped-bar.js [hunk 1]
```diff
@@ -36,9 +36,9 @@ define(function (require) {
      * @example
      * [
      *     {
-     *         "name": "2011-01",
-     *         "group": "Direct",
-     *         "value": 0
+     *         name: "2011-01",
+     *         group: "Direct",
+     *         value: 0
```
Comment (Golodhros) [inline src/charts/grouped-bar.js]: Making this consistent

File: src/charts/stacked-area.js [hunk 2]
```diff
@@ -52,9 +52,9 @@ define(function(require){
      * @example
      * [
      *     {
-     *         "date": "2011-01-05T00:00:00Z",
-     *         "name": "Direct",
-     *         "value": 0
+     *         date: "2011-01-05T00:00:00Z",
+     *         name: "Direct",
+     *         value: 0
```
Comment (Golodhros) [inline src/charts/stacked-area.js]: Same here

File: webpack.constants.js [hunk 3]
```diff
@@ -2,38 +2,40 @@ const path = require('path');
...
+    'demo-heatmap': './demos/src/demo-heatmap.js',
     'demo-kitchen-sink': './demos/src/demo-kitchen-sink.js',
-    'demo-legend': './demos/src/demo-legend.js'
+    'demo-legend': './demos/src/demo-legend.js',
+    'demo-line': './demos/src/demo-line.js',
+    'demo-scatter-plot': './demos/src/demo-scatter-plot.js',
+    'demo-stacked-area': './demos/src/demo-stacked-area.js',
+    'demo-stacked-bar': './demos/src/demo-stacked-bar.js',
+    'demo-step': './demos/src/demo-step.js',
+    'demo-sparkline': './demos/src/demo-sparkline.js'
```
Comment (Golodhros) [inline webpack.constants.js]: Ordering alphabetically

File: src/charts/heatmap.js [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+            hoursHuman = [
+                '00h', '01h', '02h', '03h', '04h', '05h', '06h', '07h', '08h',
+                '09h', '10h', '11h', '12h', '13h', '14h', '15h', '16h', '17h',
+                '18h', '19h', '20h', '21h', '22h', '23h'
+            ],
+            hourLabelHeight = 20,
+
+            getValue = ({value}) => value,
+
+            toremove;
```
Comment (dalerasrorov-eb) [inline src/charts/heatmap.js]: what does this variable represent? Is this `toRemove`? maybe a different naming could help clear this out

File: src/charts/heatmap.js [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+                .attr('width', boxSize)
+                .attr('height', boxSize)
+                .attr('x', ({hour}) => hour * boxSize)
+                .attr('y', ({day}) => day * boxSize)
+                .style('opacity', boxInitialOpacity)
+                .style('fill', boxInitialColor)
+                .style('stroke', boxBorderColor)
+                .style('stroke-width', boxBorderSize)
+                .transition()
+                    .duration(animationDuration)
```
Comment (dalerasrorov-eb) [inline src/charts/heatmap.js]: Is spacing here correct? Doesn't need two spaces instead of a tab?

File: src/charts/heatmap.js [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+         */
+        function drawHourLabels() {
+            let hourLabelsGroup = svg.select('.hour-labels-group');
+
+            hourLabels = svg.select('.hour-labels-group').selectAll('.hour-label')
+                .data(hoursHuman);
+
+            hourLabels.enter()
+              .append('text')
+                .text((label) => label)
```
Comment (dalerasrorov-eb) [inline src/charts/heatmap.js]: I think it can be just `.text(label)`

File: src/charts/heatmap.js [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+            let dayLabelsGroup = svg.select('.day-labels-group');
+
+            dayLabels = svg.select('.day-labels-group').selectAll('.day-label')
+                .data(daysHuman);
+
+            dayLabels.enter()
+              .append('text')
+                .text((label) => label)
+                .attr('x', 0)
+                .attr('y', (d, i) => i * boxSize)
```
Comment (dalerasrorov-eb) [inline src/charts/heatmap.js]: I see that there is multiple instances where `x * boxSize` where `x` is anything. Could be useful to create a private function `withBoxSize = (x) => x * boxSize` and then use like `.attr('y', (d, i) => withBoxSize(i))`. Doesn't have to be that way though, just a thought.

File: src/charts/heatmap.js (hunk 4 above)
Comment (Golodhros) [reply src/charts/heatmap.js]: Oops! Great Catch!

File: src/charts/heatmap.js (hunk 5 above)
Comment (Golodhros) [reply src/charts/heatmap.js]: Well, it seems we are not being consistent with this. I have seen no spaces, two spaces and a tab.

File: src/charts/heatmap.js (hunk 6 above)
Comment (Golodhros) [reply src/charts/heatmap.js]: Hmmm... `label` is not a variable at this point.
This came from:
  .text((d) => d)

File: src/charts/heatmap.js (hunk 7 above)
Comment (Golodhros) [reply src/charts/heatmap.js]: Yeah, that would make sense if we keep the functionality as it is right now.

Comment (Golodhros) [reply src/charts/heatmap.js]: However, I think that with the next steps, the boxes will stop being squares and start being rectangles.

**Pull request merged**